### PR TITLE
✨ [Observable] Make observable safe to remove handler while firing.

### PR DESCRIPTION
**Pull request by mszylkowski**
We need observables to remove handlers safely while firing.

If theres multiple handlers that in the function that remove themselves, only half of the handlers will fire because the iteration doesn't work well with removing items from the iteration.

```js
const observable = new Observable();
for (let i = 0; i < 100; i++) {
    const remove = observable.add(() => {
        console.log(i);
        remove();
    })
}
observable.fire();
```

This will fire only half of the observers because the `fire` iteration will remove the items as it goes (simulated run):

```js
// This skips 1/2 of the handlers
for (const handler of this.handlers_) {
    handler();
    this.handlers_.remove(handler);
}
```

is the same as 

```js
// This skips 1/2 of the handlers
for (let i = 0; i < this.handlers_.length; i++) {
    handler();
    this.handlers_.remove(handler);
}
```

beacuse after each iteration the item that's supposed to go next is at the index `i` so it gets skipped.

### CODE REVIEW

File: src/core/data-structures/observable.js [hunk 1]
```diff
@@ -58,9 +69,18 @@ export class Observable {
...
     }
+    this.iterating_ = true;
     for (const handler of this.handlers_) {
       handler(opt_event);
     }
+    this.iterating_ = false;
+    for (const handler of this.handlersToRemove_) {
+      removeItem(this.handlers_, handler);
+    }
+    if (this.handlersToRemove_.size) {
```
Comment (jridgewell) [inline src/core/data-structures/observable.js:80]: Old Code?

**Review by jridgewell — COMMENTED** [review]
This approach would be good with me. Mind splitting this from the other PR?

File: src/core/data-structures/observable.js [hunk 2]
```diff
@@ -58,9 +69,18 @@ export class Observable {
     if (!this.handlers_) {
       return;
     }
+    this.iterating_ = true;
     for (const handler of this.handlers_) {
       handler(opt_event);
     }
+    this.iterating_ = false;
```
[the remark this answers is not in the document]
Comment (mszylkowski) [reply src/core/data-structures/observable.js:76]: Do you mind explaining more? Is it that it won't work if it fires the observer again inside a fire handler?

**Review by jridgewell — APPROVED** [review]

**Comment by mszylkowski** [comment]
cc @ampproject/wg-stories

File: src/core/data-structures/observable.js [hunk 3]
```diff
@@ -69,15 +58,9 @@ export class Observable {
     if (!this.handlers_) {
       return;
     }
-    this.iteratingDepth_++;
-    for (const handler of this.handlers_) {
+    for (const handler of this.handlers_.slice()) {
```
Comment (jridgewell) [inline src/core/data-structures/observable.js:61]: Nit: Add a comment explaining.

**Review by jridgewell — APPROVED** [review]

**Review by gmajoulet — APPROVED** [review]

**Pull request merged**